daemon/api_sideload_n_try.go

```diff
@@ -207,6 +208,100 @@ out:
 	return AsyncResponse(nil, chg.ID())
 }
 
+// readForm returns a Form populated with values (for non-file parts) and file headers (for file
+// parts). The file headers contain the original file name and a path to the persisted file in
+// dirs.SnapDirBlob. Errors are *apiError.
+func readForm(reader *multipart.Reader) (form *Form, err error) {
+	maxMemory := int64(maxReadBuflen)
+	form = &Form{
+		Value:      make(map[string][]string),
+		FileHeader: make(map[string][]*FileHeader),
+	}
+
+	// clean up if we're failing the request
+	defer func() {
+		if err != nil {
+			form.RemoveAll()
+		}
+	}()
+
+	for {
+		part, err := reader.NextPart()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, BadRequest("cannot read POST form: %v", err)
+
+		}
+
+		name := part.FormName()
+		if name == "" {
+			continue
+		}
+
+		filename := part.FileName()
+		if filename == "" {
+			// non-file parts are kept in memory
+			buf := &bytes.Buffer{}
+
+			// copy one byte more than the max so we know if it exceeds the limit
+			n, err := io.CopyN(buf, part, maxMemory+1)
+			if err != nil && !errors.Is(err, io.EOF) {
+				return nil, InternalError("cannot read form data: %v", err)
+			}
+
+			maxMemory -= n
+			if maxMemory < 0 {
+				return nil, InternalError("cannot read form data: exceeds memory limit")
+			}
+
+			form.Value[name] = append(form.Value[name], buf.String())
```

Comment:
If using SnapUploadForm then we could complain if name != "snap" or if we already received a snap.

Reply:
I considered doing this since we have the info we need here but I thought it was better to keep this function just concerned with deserializing, without adding any snap logic, and then let the caller make additional checks. I think it'll make it easier to extend later (like when adding multiple snaps) and also makes it easier to reason about imo